torchrl/objectives/iql.py

```diff
@@ -357,35 +361,37 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
         else:
             tensordict_reshape = tensordict
 
-        device = self.device
-        td_device = tensordict_reshape.to(device)
-
-        loss_actor = self._loss_actor(td_device)
-        loss_qvalue, priority = self._loss_qvalue(td_device)
-        loss_value = self._loss_value(td_device)
+        loss_actor, metadata = self.actor_loss(tensordict_reshape)
+        loss_qvalue, metadata_qvalue = self.qvalue_loss(tensordict_reshape)
+        loss_value, metadata_value = self.value_loss(tensordict_reshape)
+        metadata.update(**metadata_qvalue, **metadata_value)
 
-        tensordict_reshape.set(self.tensor_keys.priority, priority)
         if (loss_actor.shape != loss_qvalue.shape) or (
             loss_value is not None and loss_actor.shape != loss_value.shape
         ):
             raise RuntimeError(
                 f"Losses shape mismatch: {loss_actor.shape}, {loss_qvalue.shape} and {loss_value.shape}"
             )
+        tensordict_reshape.set(
+            self.tensor_keys.priority, metadata.pop("td_error").detach().max(0)[0]
+        )
         if shape:
             tensordict.update(tensordict_reshape.view(shape))
         out = {
             "loss_actor": loss_actor.mean(),
             "loss_qvalue": loss_qvalue.mean(),
             "loss_value": loss_value.mean(),
-            "entropy": -td_device.get(self.tensor_keys.log_prob).mean().detach(),
+            "entropy": -tensordict_reshape.get(self.tensor_keys.log_prob)
+            .mean()
+            .detach(),
```

Comment:
I'd rather pre-compute that before putting it in the dict to avoid overloaded dict construction